lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    ch = {
+      contextPreviewExists: false
+
+      removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while editor.session.getContext(i - 1) == "equation"
+          i -= 1
+        start = i
+        while editor.session.getContext(i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
```

Comment:
It won't work, I tried. The thing is, we pass `ch.handleCurrentContext` as a callback function for Ace event, so it changes what `this` is. That's actually the reason why I named the whole object `ch` instead of `contextHandler`.